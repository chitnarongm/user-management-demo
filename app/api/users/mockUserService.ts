const fs = require("fs");
import { PrivateUser, UserData } from "@/typings/model";
import { genSaltSync, hashSync } from "bcrypt";
import { v4 as uuidv4 } from "uuid";

let users: PrivateUser[] = require("../../../data/users.json");

const saveData = () => {
  fs.writeFileSync("./data/users.json", JSON.stringify(users, null, 4));
};

const hashPasswordByName = (name: string) => {
  const salt = genSaltSync(10);
  return hashSync(`${name}pass`, salt);
};

const getAll = (limit: number, offset: number) => {
  const total = users.length;
  const start = Math.min(total - 1, offset);
  const end = Math.min(total, offset + limit);

  return {
    users: users.slice(start, end) as UserData[],
    total,
    limit,
    offset,
  };
};

const create = (user: UserData) => {
  const newUser: PrivateUser = {
    ...user,
    id: uuidv4(),
    password: hashPasswordByName(user.name),
  };

  users.unshift(newUser);

  saveData();
};

const find = (email: string): PrivateUser | undefined => {
  const tagetUser = users.find((item) => item.email === email);
  return tagetUser;
};

const update = (updatedUser: UserData) => {
  const tagetUser = users.find((item) => item.id === updatedUser.id);

  const newUser: PrivateUser = {
    ...updatedUser,
    password: hashPasswordByName(updatedUser.name),
  };

  if (tagetUser) {
    Object.assign(tagetUser, newUser);
    saveData();
  }
};

const remove = (id: string) => {
  users = users.filter((user) => user.id !== id);
  saveData();
};

export const userService = {
  getAll,
  create,
  update,
  remove,
  find,
};
