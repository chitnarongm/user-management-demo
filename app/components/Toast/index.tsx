"use client";
import styled from "@emotion/styled";
import CloseIcon from "@mui/icons-material/Close";
import { Box } from "@mui/material";
import { closeSnackbar, enqueueSnackbar, OptionsObject, SnackbarKey } from "notistack";

const CloseButton = styled(CloseIcon)({
  cursor: "pointer",
});

const action = (snackbarId: SnackbarKey) => {
  return (
    <Box role="toast-alert-close-button" onClick={() => closeSnackbar(snackbarId)}>
      <CloseButton />
    </Box>
  );
};

const options: OptionsObject = {
  variant: "error",
  action,
  autoHideDuration: 3000,
};

export const openToast = (error: any) => {
  enqueueSnackbar(<Box role="toast-alert">{error}</Box>, options);
};
