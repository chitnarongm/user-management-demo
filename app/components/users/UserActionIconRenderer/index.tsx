"use client";

import { UserData } from "@/typings/model";
import EditIcon from "@mui/icons-material/Edit";
import { IconButton } from "@mui/material";
import { GridRenderCellParams } from "@mui/x-data-grid";
import { FunctionComponent, useState } from "react";
import UserActionDialog from "../UserActionDialog";

export const UserActionIconRenderer: FunctionComponent<GridRenderCellParams<UserData>> = ({ id, field, row }) => {
  const [isShowEditDialog, setIsShowEditDialog] = useState(false);
  const openEditDialog = () => {
    setIsShowEditDialog(true);
  };

  const closeEditDialog = () => {
    setIsShowEditDialog(false);
  };

  return (
    <>
      <IconButton data-testid={`edit-icon-${field}-${id}`} onClick={openEditDialog}>
        <EditIcon />
      </IconButton>
      <UserActionDialog
        isOpen={isShowEditDialog}
        onClose={closeEditDialog}
        userData={row}
        dialogTitle="Edit User"
        confirmButtonText="Update"
        mode="edit"
        isDeleteAllowed
      />
    </>
  );
};

export default UserActionIconRenderer;
