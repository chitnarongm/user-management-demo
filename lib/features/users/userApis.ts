import { LogInForm, UserData, UserResponse } from "@/typings/model";
import axios from "axios";

export const userAxios = axios.create();

export const getUsersList = (offset: number, limit: number) => {
  return userAxios.get<UserResponse>("/api/users", {
    params: { limit, offset },
  });
};

export const putUserCreate = (userData: UserData) => {
  return userAxios.put("/api/users/create", userData);
};

export const patchUserUpdate = (userData: UserData) => {
  return userAxios.patch("/api/users/update", userData);
};

export const deleteUserDelete = (id: string) => {
  return userAxios.delete(`/api/users/delete/${id}`);
};

export const postLogin = (loginData?: LogInForm) => {
  return userAxios.post("http://localhost:3000/api/login", loginData);
};
