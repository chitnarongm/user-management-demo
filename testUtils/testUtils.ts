import { UserResponse } from "@/typings/model";

export const mockUsers = (length: number, limit: number, offset: number): UserResponse => {
  return {
    users: Array.from(Array(length)).map((item, index) => {
      return {
        id: index.toString(),
        name: `user${index}`,
        email: `user${index}@mail.com`,
        role: "user",
      };
    }),
    limit,
    offset,
    total: length,
  };
};
