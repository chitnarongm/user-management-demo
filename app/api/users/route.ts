// import type { NextApiRequest, NextApiResponse } from "next";
import { NextRequest } from "next/server";
import { userService } from "./mockUserService";

export function GET(req: NextRequest) {
  const params = req.nextUrl.searchParams;
  return Response.json(userService.getAll(Number(params.get("limit")), Number(params.get("offset"))), { status: 200 });
}
