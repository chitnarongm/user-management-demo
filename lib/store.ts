import type { Action, ThunkAction } from "@reduxjs/toolkit";
import { combineSlices, configureStore } from "@reduxjs/toolkit";
import { userSlice } from "./features/users/userSlice";

export const rootReducer = combineSlices(userSlice);
export type RootState = ReturnType<typeof rootReducer>;

export const makeStore = (preloadedState?: Partial<RootState>) => {
  return configureStore({
    reducer: rootReducer,
    preloadedState,
    middleware: (getDefaultMiddleware) => {
      return getDefaultMiddleware();
    },
  });
};

export type AppStore = ReturnType<typeof makeStore>;
export type AppDispatch = AppStore["dispatch"];
export type AppThunk<ThunkReturnType = void> = ThunkAction<ThunkReturnType, RootState, unknown, Action>;
