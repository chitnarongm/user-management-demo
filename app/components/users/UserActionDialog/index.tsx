"use client";
import { createUser, deleteUser, editUser, getUsers, selectUsersOffset } from "@/lib/features/users/userSlice";
import { useAppDispatch, useAppSelector } from "@/lib/hooks";
import { UserData } from "@/typings/model";
import { ROLES } from "@/utils/constants";
import { EMAIL_REGEX } from "@/utils/regex";
import { yupResolver } from "@hookform/resolvers/yup";
import CloseIcon from "@mui/icons-material/Close";
import {
  Button,
  Dialog,
  DialogActions,
  DialogContent,
  DialogTitle,
  FormControl,
  FormHelperText,
  IconButton,
  InputLabel,
  MenuItem,
  OutlinedInput,
  Select,
  SelectChangeEvent,
} from "@mui/material";
import { FunctionComponent, useEffect } from "react";
import { Controller, ControllerRenderProps, FormProvider, useForm } from "react-hook-form";
import { v4 as uuidv4 } from "uuid";
import * as yup from "yup";
import { openToast } from "../../Toast";

export interface UserActionDialogProps {
  isOpen: boolean;
  onClose: () => void;
  dialogTitle: string;
  confirmButtonText: string;
  mode?: "create" | "edit";
  userData?: UserData;
  isDeleteAllowed?: boolean;
}

const createUserSchema = yup.object<UserData>().shape({
  id: yup.string().required(),
  name: yup.string().required(),
  email: yup.string().required().matches(EMAIL_REGEX, "Invalid Email Format."),
  role: yup.string().required(),
});

const initialCreateUserForm: UserData = {
  id: uuidv4(),
  name: "",
  email: "",
  role: "",
};

const UserActionDialog: FunctionComponent<UserActionDialogProps> = ({
  isOpen,
  onClose,
  mode = "create",
  userData,
  dialogTitle,
  confirmButtonText,
  isDeleteAllowed = false,
}) => {
  const formMethods = useForm<UserData>({
    reValidateMode: "onSubmit",
    defaultValues: initialCreateUserForm,
    resolver: yupResolver(createUserSchema),
  });

  const dispatch = useAppDispatch();
  const offset = useAppSelector(selectUsersOffset);

  const { handleSubmit, control, unregister, reset } = formMethods;

  const handleDialogClose = () => {
    unregister("name");
    unregister("email");
    unregister("role");

    onClose();
  };

  const handleSubmitAction = async (user: UserData) => {
    switch (mode) {
      case "create":
        const createdResult = await dispatch(createUser(user));
        if (createdResult.meta.requestStatus === "fulfilled") {
          dispatch(getUsers(0));
          onClose();
        } else if (createdResult.meta.requestStatus === "rejected") {
          openToast("Create User Failed.");
        }
        break;
      case "edit":
        const editedResult = await dispatch(editUser(user));
        if (editedResult.meta.requestStatus === "fulfilled") {
          dispatch(getUsers(offset));
          onClose();
        } else if (editedResult.meta.requestStatus === "rejected") {
          openToast("Edit User Failed.");
        }
        break;
    }
  };

  const handleDeleteUser = async () => {
    if (userData) {
      const deletedResult = await dispatch(deleteUser(userData.id));
      if (deletedResult.meta.requestStatus === "fulfilled") {
        dispatch(getUsers(offset));
        onClose();
      } else if (deletedResult.meta.requestStatus === "rejected") {
        openToast("Delete User Failed.");
      }
    }
  };

  const handleRoleChange = (event: SelectChangeEvent<any>, field: ControllerRenderProps<UserData, "role">) => {
    field.onChange(event);
  };

  const prefillValue = () => {
    if (userData) {
      reset({
        id: userData.id,
        name: userData.name,
        email: userData.email,
        role: userData.role,
      });
    } else {
      reset(initialCreateUserForm);
    }
  };

  useEffect(() => {
    prefillValue();
  }, [isOpen]);

  return (
    <Dialog data-testid={`user-${mode}-dialog`} open={isOpen} onClose={handleDialogClose} fullWidth maxWidth="sm">
      <FormProvider {...formMethods}>
        <form onSubmit={handleSubmit(handleSubmitAction)}>
          <DialogTitle data-testid={`user-${mode}-title`} display="flex" justifyContent="space-between">
            {dialogTitle}
            <IconButton data-testid={`user-${mode}-close-button`} onClick={handleDialogClose}>
              <CloseIcon />
            </IconButton>
          </DialogTitle>
          <DialogContent>
            <Controller
              name="name"
              control={control}
              render={({ field, fieldState: { error } }) => (
                <FormControl margin="dense" fullWidth variant="outlined">
                  <InputLabel data-testid={`user-${mode}-name-label`} htmlFor="outlined-adornment-name">
                    Name
                  </InputLabel>
                  <OutlinedInput
                    id="outlined-adornment-name"
                    label="Name"
                    inputProps={{
                      ...field,
                      "data-testid": `user-${mode}-name-input`,
                    }}
                  />
                  <FormHelperText data-testid={`user-${mode}-name-error`} error={Boolean(error?.message)}>
                    {error?.message}
                  </FormHelperText>
                </FormControl>
              )}
            />
            <Controller
              name="email"
              control={control}
              render={({ field, fieldState: { error } }) => (
                <FormControl margin="dense" fullWidth variant="outlined">
                  <InputLabel data-testid={`user-${mode}-email-label`} htmlFor="outlined-adornment-email">
                    Email
                  </InputLabel>
                  <OutlinedInput
                    id="outlined-adornment-email"
                    label="Email"
                    inputProps={{
                      ...field,
                      "data-testid": `user-${mode}-email-input`,
                    }}
                  />
                  <FormHelperText data-testid={`user-${mode}-email-error`} error={Boolean(error?.message)}>
                    {error?.message}
                  </FormHelperText>
                </FormControl>
              )}
            />
            <Controller
              name="role"
              control={control}
              render={({ field, fieldState: { error } }) => (
                <FormControl margin="dense" fullWidth variant="outlined">
                  <InputLabel data-testid={`user-${mode}-role-label`} id="demo-simple-select-label">
                    Role
                  </InputLabel>
                  <Select
                    labelId="demo-simple-select-label"
                    id="demo-simple-select"
                    label="Role"
                    onChange={(event) => handleRoleChange(event, field)}
                    // @ts-ignore
                    SelectDisplayProps={{ "data-testid": `user-${mode}-role-select` }}
                    inputProps={{
                      ...field,
                      "data-testid": `user-${mode}-role-input`,
                    }}
                  >
                    {Object.entries(ROLES).map(([value, displayValue], index) => (
                      <MenuItem
                        role={`user-${mode}-role-select-item`}
                        data-testid={`user-${mode}-role-select-item-${index}`}
                        key={value}
                        value={value}
                      >
                        {displayValue}
                      </MenuItem>
                    ))}
                  </Select>
                  <FormHelperText data-testid={`user-${mode}-role-error`} error={Boolean(error?.message)}>
                    {error?.message}
                  </FormHelperText>
                </FormControl>
              )}
            />
          </DialogContent>
          <DialogActions>
            {isDeleteAllowed && (
              <Button
                data-testid={`user-${mode}-delete-button`}
                variant="contained"
                color="error"
                onClick={handleDeleteUser}
              >
                Delete
              </Button>
            )}
            <Button data-testid={`user-${mode}-confirm-button`} variant="contained" type="submit">
              {confirmButtonText}
            </Button>
          </DialogActions>
        </form>
      </FormProvider>
    </Dialog>
  );
};

export default UserActionDialog;
