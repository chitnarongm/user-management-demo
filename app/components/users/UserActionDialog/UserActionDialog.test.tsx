import { deleteUserDelete, patchUserUpdate, putUserCreate } from "@/lib/features/users/userApis";
import { makeStore } from "@/lib/store";
import { renderWithProviders } from "@/testUtils/mockReduxStore";
import { screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import UserActionDialog, { UserActionDialogProps } from ".";
import { openToast } from "../../Toast";

jest.mock("../../../../lib/features/users/userApis");
jest.mock("../../Toast");

describe("UserActionDialog", () => {
  let props: UserActionDialogProps;
  let store = makeStore();

  beforeEach(() => {
    props = {
      isOpen: true,
      onClose: jest.fn(),
      dialogTitle: "Dialog Title",
      confirmButtonText: "Confirm;",
    };
  });

  const renderUserActionDialog = () => {
    return renderWithProviders(<UserActionDialog {...props} />, { store });
  };

  it("should render correctly", () => {
    const { baseElement } = renderUserActionDialog();

    expect(baseElement).toMatchSnapshot();
  });

  describe("Create User", () => {
    beforeEach(() => {
      props.dialogTitle = "Create User";
      props.confirmButtonText = "Create";
    });

    it("should display information correctly", async () => {
      renderUserActionDialog();

      expect(screen.getByTestId("user-create-dialog")).toBeVisible();
      expect(screen.getByTestId("user-create-title")).toHaveTextContent("Create User");
      expect(screen.getByTestId("user-create-close-button")).toBeVisible();

      expect(screen.getByTestId("user-create-name-label")).toHaveTextContent("Name");
      expect(screen.getByTestId("user-create-name-input")).toHaveDisplayValue("");
      expect(screen.getByTestId("user-create-name-error")).toBeEmptyDOMElement();

      expect(screen.getByTestId("user-create-email-label")).toHaveTextContent("Email");
      expect(screen.getByTestId("user-create-email-input")).toHaveDisplayValue("");
      expect(screen.getByTestId("user-create-email-error")).toBeEmptyDOMElement();

      expect(screen.getByTestId("user-create-role-label")).toHaveTextContent("Role");
      expect(screen.getByTestId("user-create-role-input")).toHaveDisplayValue("");
      expect(screen.getByTestId("user-create-role-error")).toBeEmptyDOMElement();

      await userEvent.click(screen.getByTestId("user-create-role-select"));

      expect(screen.getByTestId("user-create-role-select-item-0")).toHaveTextContent("User");
      expect(screen.getByTestId("user-create-role-select-item-1")).toHaveTextContent("Moderator");
      expect(screen.getByTestId("user-create-role-select-item-2")).toHaveTextContent("Admin");

      expect(screen.queryByTestId("user-create-delete-button")).not.toBeInTheDocument();
      expect(screen.getByTestId("user-create-confirm-button")).toHaveTextContent("Create");
    });

    it("should create user successfully when click create button", async () => {
      renderUserActionDialog();

      await userEvent.type(screen.getByTestId("user-create-name-input"), "username");
      await userEvent.type(screen.getByTestId("user-create-email-input"), "mock@email.com");
      await userEvent.click(screen.getByTestId("user-create-role-select"));
      await userEvent.click(screen.getByTestId("user-create-role-select-item-0"));

      await userEvent.click(screen.getByTestId("user-create-confirm-button"));

      expect(putUserCreate).toHaveBeenCalledWith({
        email: "mock@email.com",
        id: "mockuuidv4",
        name: "username",
        role: "user",
      });

      expect(props.onClose).toHaveBeenCalled();
      expect(openToast).not.toHaveBeenCalled();
    });

    it("should not close dialog when create user failed", async () => {
      jest.mocked(putUserCreate).mockRejectedValue({ message: "Create User Error" });

      renderUserActionDialog();

      await userEvent.type(screen.getByTestId("user-create-name-input"), "username");
      await userEvent.type(screen.getByTestId("user-create-email-input"), "mock@email.com");
      await userEvent.click(screen.getByTestId("user-create-role-select"));
      await userEvent.click(screen.getByTestId("user-create-role-select-item-0"));

      await userEvent.click(screen.getByTestId("user-create-confirm-button"));

      expect(putUserCreate).toHaveBeenCalledWith({
        email: "mock@email.com",
        id: "mockuuidv4",
        name: "username",
        role: "user",
      });

      expect(props.onClose).not.toHaveBeenCalled();
      expect(openToast).toHaveBeenCalledWith("Create User Failed.");
    });

    it("should call onClose prop when click close button", async () => {
      renderUserActionDialog();

      await userEvent.click(screen.getByTestId("user-create-close-button"));

      expect(props.onClose).toHaveBeenCalled();
    });
  });

  describe("Update and Delete User", () => {
    beforeEach(() => {
      props.mode = "edit";
      props.dialogTitle = "Update User";
      props.confirmButtonText = "Update";
      props.userData = {
        id: "mockprefilleduserid",
        name: "prefilledusername",
        email: "prefilled@email.com",
        role: "moderator",
      };
    });

    it("should display prefilled information correctly", async () => {
      renderUserActionDialog();

      expect(screen.getByTestId("user-edit-dialog")).toBeVisible();
      expect(screen.getByTestId("user-edit-title")).toHaveTextContent("Update User");
      expect(screen.getByTestId("user-edit-close-button")).toBeVisible();

      expect(screen.getByTestId("user-edit-name-label")).toHaveTextContent("Name");
      expect(screen.getByTestId("user-edit-name-input")).toHaveDisplayValue("prefilledusername");
      expect(screen.getByTestId("user-edit-name-error")).toBeEmptyDOMElement();

      expect(screen.getByTestId("user-edit-email-label")).toHaveTextContent("Email");
      expect(screen.getByTestId("user-edit-email-input")).toHaveDisplayValue("prefilled@email.com");
      expect(screen.getByTestId("user-edit-email-error")).toBeEmptyDOMElement();

      expect(screen.getByTestId("user-edit-role-label")).toHaveTextContent("Role");
      expect(screen.getByTestId("user-edit-role-select")).toHaveTextContent("Moderator");
      expect(screen.getByTestId("user-edit-role-error")).toBeEmptyDOMElement();

      await userEvent.click(screen.getByTestId("user-edit-role-select"));

      expect(screen.getByTestId("user-edit-role-select-item-0")).toHaveTextContent("User");
      expect(screen.getByTestId("user-edit-role-select-item-1")).toHaveTextContent("Moderator");
      expect(screen.getByTestId("user-edit-role-select-item-2")).toHaveTextContent("Admin");

      expect(screen.queryByTestId("user-edit-delete-button")).not.toBeInTheDocument();
      expect(screen.getByTestId("user-edit-confirm-button")).toHaveTextContent("Update");
    });

    it("should update user successfully when click update button", async () => {
      renderUserActionDialog();

      await userEvent.click(screen.getByTestId("user-edit-confirm-button"));

      expect(patchUserUpdate).toHaveBeenCalledWith({
        email: "prefilled@email.com",
        id: "mockprefilleduserid",
        name: "prefilledusername",
        role: "moderator",
      });

      expect(props.onClose).toHaveBeenCalled();
      expect(openToast).not.toHaveBeenCalled();
    });

    it("should not close dialog when update user failed", async () => {
      jest.mocked(patchUserUpdate).mockRejectedValue({ message: "Update User Error" });

      renderUserActionDialog();

      await userEvent.click(screen.getByTestId("user-edit-confirm-button"));

      expect(patchUserUpdate).toHaveBeenCalledWith({
        email: "prefilled@email.com",
        id: "mockprefilleduserid",
        name: "prefilledusername",
        role: "moderator",
      });

      expect(props.onClose).not.toHaveBeenCalled();
      expect(openToast).toHaveBeenCalledWith("Edit User Failed.");
    });

    it("should display delete button when isDeleteAllowed prop is true and delete user successfully when click delete button", async () => {
      props.isDeleteAllowed = true;

      renderUserActionDialog();

      expect(screen.getByTestId("user-edit-delete-button")).toHaveTextContent("Delete");

      await userEvent.click(screen.getByTestId("user-edit-delete-button"));

      expect(deleteUserDelete).toHaveBeenCalledWith("mockprefilleduserid");
      expect(props.onClose).toHaveBeenCalled();
      expect(openToast).not.toHaveBeenCalled();
    });

    it("should not close dialog when delete user failed", async () => {
      jest.mocked(deleteUserDelete).mockRejectedValue({ message: "Delete User Error" });
      props.isDeleteAllowed = true;

      renderUserActionDialog();

      expect(screen.getByTestId("user-edit-delete-button")).toHaveTextContent("Delete");

      await userEvent.click(screen.getByTestId("user-edit-delete-button"));

      expect(deleteUserDelete).toHaveBeenCalledWith("mockprefilleduserid");
      expect(props.onClose).not.toHaveBeenCalled();
      expect(openToast).toHaveBeenCalledWith("Delete User Failed.");
    });
  });

  describe("Validation", () => {
    it("should validate required fields correctly", async () => {
      renderUserActionDialog();

      await userEvent.click(screen.getByTestId("user-create-confirm-button"));

      expect(screen.getByTestId("user-create-name-error")).toHaveTextContent("name is a required field");
      expect(screen.getByTestId("user-create-email-error")).toHaveTextContent("email is a required field");
      expect(screen.getByTestId("user-create-role-error")).toHaveTextContent("role is a required field");
    });

    it("should validate email pattern correctly", async () => {
      renderUserActionDialog();

      await userEvent.type(screen.getByTestId("user-create-email-input"), "invalid..email..pattern");

      await userEvent.click(screen.getByTestId("user-create-confirm-button"));

      expect(screen.getByTestId("user-create-email-error")).toHaveTextContent("Invalid Email Format.");
    });
  });
});
