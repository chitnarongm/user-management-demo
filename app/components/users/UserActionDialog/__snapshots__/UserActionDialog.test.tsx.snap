// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserActionDialog should render correctly 1`] = `
<body
  style="padding-right: 1024px; overflow: hidden;"
>
  <div
    aria-hidden="true"
  />
  <div
    class="MuiDialog-root MuiModal-root css-zw3mfo-MuiModal-root-MuiDialog-root"
    data-testid="user-create-dialog"
    role="presentation"
  >
    <div
      aria-hidden="true"
      class="MuiBackdrop-root MuiModal-backdrop css-yiavyu-MuiBackdrop-root-MuiDialog-backdrop"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    />
    <div
      data-testid="sentinelStart"
      tabindex="0"
    />
    <div
      class="MuiDialog-container MuiDialog-scrollPaper css-hz1bth-MuiDialog-container"
      role="presentation"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
      tabindex="-1"
    >
      <div
        aria-labelledby=":r0:"
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation24 MuiDialog-paper MuiDialog-paperScrollPaper MuiDialog-paperWidthSm MuiDialog-paperFullWidth css-tlc64q-MuiPaper-root-MuiDialog-paper"
        role="dialog"
      >
        <form>
          <h2
            class="MuiTypography-root MuiTypography-h6 MuiDialogTitle-root css-yipdbk-MuiTypography-root-MuiDialogTitle-root"
            data-testid="user-create-title"
            id=":r0:"
          >
            Dialog Title
            <button
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
              data-testid="user-create-close-button"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="CloseIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </h2>
          <div
            class="MuiDialogContent-root css-ypiqx9-MuiDialogContent-root"
          >
            <div
              class="MuiFormControl-root MuiFormControl-marginDense MuiFormControl-fullWidth css-zevgu-MuiFormControl-root"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="false"
                data-testid="user-create-name-label"
                for="outlined-adornment-name"
              >
                Name
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                  data-testid="user-create-name-input"
                  id="outlined-adornment-name"
                  name="name"
                  type="text"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-yjsfm1"
                  >
                    <span>
                      Name
                    </span>
                  </legend>
                </fieldset>
              </div>
              <p
                class="MuiFormHelperText-root MuiFormHelperText-sizeMedium MuiFormHelperText-contained css-1wc848c-MuiFormHelperText-root"
                data-testid="user-create-name-error"
              />
            </div>
            <div
              class="MuiFormControl-root MuiFormControl-marginDense MuiFormControl-fullWidth css-zevgu-MuiFormControl-root"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="false"
                data-testid="user-create-email-label"
                for="outlined-adornment-email"
              >
                Email
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                  data-testid="user-create-email-input"
                  id="outlined-adornment-email"
                  name="email"
                  type="text"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-yjsfm1"
                  >
                    <span>
                      Email
                    </span>
                  </legend>
                </fieldset>
              </div>
              <p
                class="MuiFormHelperText-root MuiFormHelperText-sizeMedium MuiFormHelperText-contained css-1wc848c-MuiFormHelperText-root"
                data-testid="user-create-email-error"
              />
            </div>
            <div
              class="MuiFormControl-root MuiFormControl-marginDense MuiFormControl-fullWidth css-zevgu-MuiFormControl-root"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="false"
                data-testid="user-create-role-label"
                id="demo-simple-select-label"
              >
                Role
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl  css-1yk1gt9-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
              >
                <div
                  aria-controls=":r1:"
                  aria-expanded="false"
                  aria-haspopup="listbox"
                  aria-labelledby="demo-simple-select-label demo-simple-select"
                  class="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
                  data-testid="user-create-role-select"
                  id="demo-simple-select"
                  role="combobox"
                  tabindex="0"
                >
                  <span
                    class="notranslate"
                  >
                    â€‹
                  </span>
                </div>
                <input
                  aria-hidden="true"
                  aria-invalid="false"
                  class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
                  data-testid="user-create-role-input"
                  name="role"
                  tabindex="-1"
                  value=""
                />
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
                  data-testid="ArrowDropDownIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M7 10l5 5 5-5z"
                  />
                </svg>
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-yjsfm1"
                  >
                    <span>
                      Role
                    </span>
                  </legend>
                </fieldset>
              </div>
              <p
                class="MuiFormHelperText-root MuiFormHelperText-sizeMedium MuiFormHelperText-contained css-1wc848c-MuiFormHelperText-root"
                data-testid="user-create-role-error"
              />
            </div>
          </div>
          <div
            class="MuiDialogActions-root MuiDialogActions-spacing css-knqc4i-MuiDialogActions-root"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-sghohy-MuiButtonBase-root-MuiButton-root"
              data-testid="user-create-confirm-button"
              tabindex="0"
              type="submit"
            >
              Confirm;
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </form>
      </div>
    </div>
    <div
      data-testid="sentinelEnd"
      tabindex="0"
    />
  </div>
</body>
`;
